resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala

```diff
@@ -397,6 +407,64 @@ private[spark] class KubernetesClusterSchedulerBackend(
     }
   }
 
+  /*
+ * KubernetesAllocator class watches all the executor pods associated with
+ * this SparkJob and creates new executors when it is appropriate.
+ */
+  private[spark] class KubernetesAllocator(
+    client: KubernetesClient,
+    interval: Int,
+    backend: KubernetesClusterSchedulerBackend) extends Logging {
+
+    private val scheduler = Executors.newScheduledThreadPool(1)
+    private val allocatorRunnable: Runnable = new Runnable {
+      override def run(): Unit = {
+        if (readyExecutorPods.size < totalExpectedExecutors.get * minRegisteredRatio) {
+          logInfo("Waiting for pending executors before scaling")
+          return
+        }
+
+        if (runningExecutorPods.size >= executorLimitOption.get) {
+          logInfo("Maximum allowed executor limit reached. Not scaling up further.")
+          return
+        }
+
+        EXECUTOR_MODIFICATION_LOCK.synchronized {
+          runningExecutorPods += allocateNewExecutorPod()
+          totalExpectedExecutors.set(runningExecutorPods.size)
+          logInfo(s"Requesting a new executor, total executors is now ${totalExpectedExecutors}")
+        }
+      }
+    }
+    if (interval > 0) {
+      scheduler.scheduleWithFixedDelay(allocatorRunnable, 0, interval, TimeUnit.SECONDS)
+    }
+
+    // This is a set of all ready executor pods.
+    private var readyExecutorPods = Set[String]()
+
+    client
```

Comment:
Start the watch in `start()` and keep a reference to it. Close the watch in `stop()`.